atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java

```diff
@@ -0,0 +1,256 @@
+    final var context = new RaftClusterContext(localMember.memberId(), raft);
+
+    // when
+    final Consumer<Boolean> callback = mock();
+    final var quorum = context.getVoteQuorum(callback);
+
+    quorum.succeed(new MemberId("2"));
+
+    // then
+    verify(callback).accept(true);
+  }
+
+  @Test
+  void shouldRequireJointConsensusVotes() {
+    // given
+    final var localMember = new DefaultRaftMember(new MemberId("1"), Type.ACTIVE, Instant.now());
+    final var oldRemoteMembers =
+        List.<RaftMember>of(
+            new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(new MemberId("3"), Type.ACTIVE, Instant.now()));
+    final var oldMembers =
+        Stream.concat(Stream.of(localMember), oldRemoteMembers.stream()).toList();
+    final var newRemoteMembers =
+        List.<RaftMember>of(
+            new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(new MemberId("4"), Type.ACTIVE, Instant.now()));
+    final var newMembers =
+        Stream.concat(Stream.of(localMember), newRemoteMembers.stream()).toList();
+
+    final var raft =
+        raftWithStoredConfiguration(
+            new Configuration(1, 1, Instant.now().toEpochMilli(), newMembers, oldMembers));
+    final var context = new RaftClusterContext(localMember.memberId(), raft);
+
+    // when
+    final Consumer<Boolean> callback = mock();
+    final var quorum = context.getVoteQuorum(callback);
+
+    // then
+    quorum.succeed(new MemberId("4"));
+    verifyNoInteractions(callback);
+
+    quorum.succeed(new MemberId("2"));
+    verify(callback).accept(true);
+  }
+
+  @Test
+  void shouldCalculateQuorum() {
+    // given
+    final var localMember = new DefaultRaftMember(new MemberId("1"), Type.ACTIVE, Instant.now());
+    final var remoteMembers =
+        List.<RaftMember>of(
+            new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(new MemberId("3"), Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(new MemberId("4"), Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(new MemberId("5"), Type.ACTIVE, Instant.now()));
+    final var members = Stream.concat(Stream.of(localMember), remoteMembers.stream()).toList();
+
+    final var raft =
+        raftWithStoredConfiguration(new Configuration(1, 1, Instant.now().toEpochMilli(), members));
+    final var context = new RaftClusterContext(localMember.memberId(), raft);
+
+    // when
+    context.getMemberContext(new MemberId("2")).setMatchIndex(2);
+    context.getMemberContext(new MemberId("3")).setMatchIndex(3);
+    context.getMemberContext(new MemberId("4")).setMatchIndex(4);
+    context.getMemberContext(new MemberId("5")).setMatchIndex(5);
+
+    // then
+    assertThat(context.getQuorumFor(RaftMemberContext::getMatchIndex)).hasValue(4L);
+  }
+
+  @Test
+  void shouldCalculateQuorumDuringJointConsensus() {
```

Comment:
:wrench: : Also add test where matchIndex in oldMembers is higher than matchIndex in new members.